src/transactions.rs

```diff
@@ -1000,16 +1104,20 @@ impl WriteTransaction {
     }
 
     fn commit_inner(&mut self) -> Result<(), CommitError> {
+        // Quick-repair requires 2-phase commit
+        if self.quick_repair {
+            self.two_phase_commit = true;
+        }
+
```

Comment:
I think this is unsafe even with `Durability::Immediate`.

Suppose your application makes some commits that contain attacker-controlled data, and others that don't.  You might think you can use 2-phase commit just for the former, but that's not good enough!  The attacker's data is present in the database, so they can cause a crash at the moment when it's about to be overwritten by a new (non-attacker-controlled) page, leaving the new page pointer referring to the attacker-controlled data instead.  Since we're already assuming they can predict the database layout and the content of the data they don't control, they can forge the checksum to make it appear valid.

If you want the security of 2-phase commit, you need to use it for every commit during which the attacker could cause a crash, which is probably all of them.

Reply:
That said, I agree that 2-phase commit with `Durability::None` isn't really a meaningful combination, and I'd be okay with disallowing it.  (I'd rather do that than silently upgrade to `Durability::Immediate`, because the user presumably cares about commit speed if they're requesting `Durability::None`; we should tell them how to get that speed rather than silently doing the wrong thing.)

But my preference is to allow the combination, even though it doesn't do anything, for exactly the reason mentioned above: the right way to use 2-phase commit is to enable it on every single commit, so we may as well make that easy.  The same thing applies to quick-repair -- if you're using quick-repair, you probably want it on every commit, so it seems convenient to allow it even on the `Durability::None` commits where it has no effect.

Reply:
Oh ya, you're right! Security is hard, haha